b2/account_info.py

```diff
@@ -114,123 +141,258 @@ def clear_large_file_upload_urls(self, file_id):
-        else:
-            self.filename = file_name
-        self._lock_filename = self.filename + '.lock'
-        self._lock_timeout = internal_lock_timeout
+    def __init__(self, file_name=None):
+        self.thread_local = threading.local()
+        user_account_info_path = file_name or os.environ.get(
+            'B2_ACCOUNT_INFO', '~/.b2_account_info'
+        )
+        self.filename = file_name or os.path.expanduser(user_account_info_path)
+        self._validate_database()
+        with self._get_connection() as conn:
+            self._create_tables(conn)
+
         self._large_file_uploads = collections.defaultdict(
             list
         )  # We don't keep large file upload URLs across a reload
-        self._bucket_names_to_ids = {}  # for in-memory cache
 
-        # Lock to manage in-memory structures
+        # this lock controls access to self._large_file_uploads
         self._lock = threading.Lock()
 
+    def _validate_database(self):
+        """
+        Makes sure that the database is openable.  Removes the file if it's not.
+        """
+        # If there is no file there, that's fine.  It will get created when
+        # we connect.
+        if not os.path.exists(self.filename):
+            self._create_database()
+            return
+
+        # If we can connect to the database, and do anything, then all is good.
+        try:
+            with self._connect() as conn:
+                self._create_tables(conn)
+                return
+        except:
+            pass  # fall through to next case
+
+        # If the file contains JSON with the right stuff in it, convert from
+        # the old representation.
+        try:
+            with open(self.filename, 'rb') as f:
+                data = json.loads(f.read().decode('utf-8'))
+                keys = [
+                    'account_id', 'application_key', 'account_auth_token', 'api_url',
+                    'download_url', 'minimum_part_size', 'realm'
+                ]
+                if all(k in data for k in keys):
+                    # remove the json file
+                    os.unlink(self.filename)
+                    # create a database
+                    self._create_database()
+                    # add the data from the JSON file
+                    with self._connect() as conn:
+                        self._create_tables(conn)
+                        insert_statement = """
+                            INSERT INTO account
+                            (account_id, application_key, account_auth_token, api_url, download_url, minimum_part_size, realm)
+                            values (?, ?, ?, ?, ?, ?, ?);
+                        """
+
+                        conn.execute(insert_statement, tuple(data[k] for k in keys))
+                    # all is happy now
+                    return
+        except:
+            pass
+
+        # Remove the corrupted file and create a new database
```

Comment:
if this block will stay, it should go to the `except` clause and `else` (under `with`) to prevent fall-through due to future changes in the code above, from deleting a file.